### ISSUE TITLE
performance of Scanner - avoid copying of collections

### ISSUE THREAD

**Issue by pvojtechovsky** [INRIA/spoon#1137]
The `CtScanner' has method like this:
```java
	public void scan(Collection<? extends CtElement> elements) {
		if ((elements != null)) {
			// we use defensive copy so as to be able to change the class while scanning
			// otherwise one gets a ConcurrentModificationException
			for (CtElement e : new ArrayList<>(elements)) {
				scan(e);
			}
		}
       }
```
it always copies all visited collections of the model. I have not measured the performance penalty, so I have no idea how much faster it can be if we avoid that.

How to avoid it and still support modification of the model during scanning process?

We have to change the implementation of each model collection. We have to replace each add/remove method body which calls
```java
collection.add() or collection.remove();
```
by
```java
copy = collection.clone();
copy.add() or copy.remove();
collection=copy;
```
It will influence performance of model creation/manipulation, but I guess creation/manipulation is less often then scanning.

May be we can implement and use some special set of collections, which are aware of their Iterators and in case when iterator exists it would make a copy of internal storage automatically. That would probably does not need so many changes in the implementation of the model.

Note: I am not going to implement that, but may be somebody who needs better performance might be interested in this idea...

**Comment by msteinbeck** [comment]
Is https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArrayList.html what you are looking for?

**Comment by pvojtechovsky** [comment]
> CopyOnWriteArrayList

yes, that implementation looks like good beginning. But I have feeling like we might improve implementation of that List to be more efficient for spoon use case.
For example 
* we do not need to support multithreading
* we should support Iterator#remove/add
* we should copy the underlaying array only if there is an iteration in progress, otherwise we can modify existing array.
* ...

**Comment by msteinbeck** [comment]
> we do not need to support multithreading

As far as I know multithreading is not implemented by using `synchronized` but by the way how this List works, copy the entire array on write.

> we should support Iterator#remove/add

I don't think that this is possible. An iterator that, on one hand, iterates through a snapshot of a collection and, on the the other hand, allows operations changing the original data sounds contradictory.

How about adding a constructor to `CtScanner` taking a boolean that indicates whether a scanner is used to modify the AST. For instance:

```java
private final boolean readonly;

public CtScanner(boolean readonly) {...}

...

public void scan(Collection<? extends CtElement> elements) {
	if ((elements != null)) {
		// we use defensive copy so as to be able to change the class while scanning
		// otherwise one gets a ConcurrentModificationException
		for (CtElement e : readonly ? elements : new ArrayList<>(elements)) {
			scan(e);
		}
	}
}

```

That way third party software doesn't need to change its implementation and Spoon's internal compiler (as well as others) could utilize the enhanced API.

**Comment by pvojtechovsky** [comment]
> > we should support Iterator#remove/add
> 
> I don't think that this is possible.

It is not possible in multithreaded environment, but with single thread it might work. 

> adding a constructor to CtScanner taking a boolean that indicates whether a scanner is used to modify the AST

good idea. Very simple and will work for special cases, But it needs extra work on client's side. If we would have a clever List which is able to handle Iterator#add/remove then this extra work would not be needed.

**Comment by msteinbeck** [comment]
I performed a small test and didn't notice any speed changes when not copying the list before starting iteration.

**Comment by pvojtechovsky** [comment]
Thanks!
OK, so copying of arrays is probably fast operation in java

**Comment by monperrus** [comment]
Interesting discussion.

> I have not measured the performance penalty, so I have no idea how much faster it can be if we avoid that.

This would be a good thing to do first.

**Comment by pvojtechovsky** [comment]
Yesterday I did tests, which scans 3M model items. There was no other operation with these items. Just scanning. The profiler displayed that the array allocation takes about 20% of time. But if I tried to improve it, it started to behave strange. The code which just iterates over Collection, without it's previous copying was little bit slower. Even just calling super was slower. I will elaborate it further, but it is true that 
* array copying in java is probably very fast
* relevant measuring of performance is hard task
* the first scanning is always much slower then second - probably java precompilation or CPU caches plays role in it?

Anyway the array copying has no visible performance effect on creation of spoon model. May be it will have effect on analyzing algorithms which scans the model intensively.

**Comment by msteinbeck** [comment]
> array copying in java is probably very fast

Copying arrays in Java is a system call as indicated with the `native` keyword:

```java
public static native void arraycopy(Object src,  int  srcPos,
                                        Object dest, int destPos,
                                        int length);
```

Since this is a operation used by many programs, modern operating systems provider a very fast implementation.

**Comment by msteinbeck** [comment]
You can find some performance measurements at https://github.com/INRIA/spoon/issues/1139 showing that one of the most time consuming tasks is the `ParentExiter` used by `ContextBuilder` when exiting a node.

**Comment by pvojtechovsky** [comment]
The #1139 is about building of model. I wanted to understand how fast is the pure model scanning, because the query algorithms I am going to use in core of my future work, will use that model scanning intensively - may be it will do scanning all the time for many minutes ... then it will make sense to optimize scanning. But let's wait until it really comes. Till now the scanner is fast enough and does not need improvement for my use case.

**Comment by msteinbeck** [comment]
Is it possible for you to parallelize some of your operations? We have several scanner that we use in parallel in order to speed up the overall traversal.

**Comment by pvojtechovsky** [comment]
interesting idea, thanks.

Another think which surprised me, was that scanning of model creates helper references. Originally I though that model scanning just goes through model, but I see that it also creates types references, etc on the fly. It has two disadvantages
* you cannot use java == operator to check weather you already processed that element (in case of analyze which uses multiple scanning passes)
* it is little bit slower then returning of existing reference would be
May be it is so to minimize memory... or it is just bug?

**Comment by msteinbeck** [comment]
> you cannot use java == operator to check weather you already processed that element (in case of analyze which uses multiple scanning passes)

That's why I'm using the qualified name of a reference. `hashCode` and `equals` is much faster of Strings than on references.

**Comment by pvojtechovsky** [comment]
@msteinbeck How do  you generate qualified name on local variable reference?

**Comment by msteinbeck** [comment]
> How do you generate qualified name on local variable reference?

local variables are one of the nodes of which I don't need to generate a qualified name for. What you could do is to generate the qualified name of its declaring method, use a custom separator, and add the name of the variable by your own.

**Comment by msteinbeck** [comment]
Can we close this one? It looks like copying of collections doesn't slow down Spoon, right?

**Comment by pvojtechovsky** [comment]
We can close it. This problem exists - scanning might be faster, but it is fast enough for me for now.
If somebody wants it faster then s/he can focus on:
* copying of collections
* creation of virtual helper objects, which are created just to be visited and then removed again.
